gateway/src/api/auth_layer.rs

```diff
@@ -225,6 +289,35 @@ where
     }
 }
 
+fn extract_token_expiration(token: String) -> Result<Duration, StatusCode> {
+    let (_header, rest) = token
+        .split_once('.')
+        .ok_or(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+    let (claim, _sig) = rest
+        .split_once('.')
+        .ok_or(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+    let claim = base64::decode_config(claim, base64::URL_SAFE_NO_PAD)
+        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;
+
+    let claim: serde_json::Map<String, serde_json::Value> =
+        serde_json::from_slice(&claim).map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;
+
+    let exp = claim["exp"]
+        .as_i64()
+        .ok_or(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+    let expiration_timestamp = Utc
+        .timestamp_opt(exp, 0)
+        .single()
+        .ok_or(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+    let duration = expiration_timestamp - Utc::now();
```

Comment:
I've been thinking whether we want to subtract an extra 30 seconds or minute from the duration so that the cache expires a little bit before the token?

I'm worried about the case where a user might do a `cargo shuttle deploy` when their token cache might expire in 30 seconds or so. In this case the token will reach deployer which might take longer than 30 seconds to build the project before sending it on to provisioner which will fail saying the token is invalid.

Reply:
I've had the same thought, I agree, we should do that. We can start with 30 seconds and then test that it's enough of a buffer on monday.

Reply:
Subtracting 60 seconds for now.